src/main/java/cz/cvut/kbss/study/persistence/dao/BaseDao.java

```diff
@@ -65,129 +50,63 @@ protected List<T> findAll(EntityManager em) {
     @Override
     public void persist(T entity) {
         Objects.requireNonNull(entity);
-        final EntityManager em = entityManager();
         try {
-            em.getTransaction().begin();
-            persist(entity, em);
-            em.getTransaction().commit();
+            em.persist(entity);
         } catch (RuntimeException e) {
             LOG.error("Error when persisting entity.", e);
             throw new PersistenceException(e);
-        } finally {
-            em.close();
         }
     }
 
-    protected void persist(T entity, EntityManager em) {
-        em.persist(entity);
-    }
-
     @Override
     public void persist(Collection<T> entities) {
         Objects.requireNonNull(entities);
         if (entities.isEmpty()) {
             return;
         }
-        final EntityManager em = entityManager();
-        try {
-            em.getTransaction().begin();
-            entities.forEach(e -> persist(e, em));
-            em.getTransaction().commit();
-        } catch (RuntimeException e) {
-            LOG.error("Error when persisting entities.", e);
-            throw new PersistenceException(e);
-        } finally {
-            em.close();
-        }
+        entities.forEach(this::persist);
     }
 
     @Override
     public void update(T entity) {
         Objects.requireNonNull(entity);
-        final EntityManager em = entityManager();
         try {
-            em.getTransaction().begin();
-            update(entity, em);
-            em.getTransaction().commit();
+            em.merge(entity);
         } catch (RuntimeException e) {
             LOG.error("Error when updating entity.", e);
             throw new PersistenceException(e);
-        } finally {
-            em.close();
         }
     }
 
-    protected void update(T entity, EntityManager em) {
-        em.merge(entity);
-    }
-
     @Override
     public void remove(T entity) {
         Objects.requireNonNull(entity);
-        final EntityManager em = entityManager();
         try {
-            em.getTransaction().begin();
-            remove(entity, em);
-            em.getTransaction().commit();
+            final T toRemove = em.merge(entity);
+            assert toRemove != null;
```

Comment:
```suggestion
            Objects.requireNonNull(toRemove);
```

Reply:
`toRemove` will never be `null` (`EntityManager.merge` does not return null under any circumstances) so there is no point calling `Objects.requireNonNull`. The `assert` is mostly for documentation reasons.